### ISSUE TITLE
"Fatal: Couldn't find any working encoder" After switching to wayland.

### ISSUE THREAD

**Issue by KelvinGuy1** [LizardByte/Sunshine#1916]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your issue described in the documentation?

- [X] I have read the documentation

### Is your issue present in the nightly release?

- [X] This issue is present in the nightly release

### Describe the Bug

After moving to Wayland, when attempting to run sunshine on Ubuntu, it can't find an encoder.

### Host Operating System

Linux

### Operating System Version

Ubuntu 22.04.2 LTS

### Architecture

64 bit

### Sunshine commit or version

0.19 to latest ver and flathub build

### Package

Linux - deb + flatpak

### GPU Type

Nvidia

### GPU Model

GeForce 1660 Ti

### GPU Driver/Mesa Version

525 proprietary

### Config

```Shell
output_name = 1
sw_preset = ultrafast
origin_web_ui_allowed = pc
nv_preset = p1
sw_tune = fastdecode
min_threads = 36
qp = -2
encoder = vaapi
address_family = both
capture = wlr
av1_mode = 1
```

### Relevant log output

```shell
[encoder] -- [nvenc]
[qp] -- [-2]
[min_threads] -- [36]
[sw_tune] -- [fastdecode]
[nv_preset] -- [p1]
[origin_web_ui_allowed] -- [pc]
[capture] -- [nvfbc]
[sw_preset] -- [ultrafast]
[output_name] -- [1]
Warning: Unrecognized configurable option [nv_preset]
[2023:12:16:03:03:11]: Info: Sunshine version: 0.21.0.4c984cacb54c09ee954e8855b6341e699bbfd535
[2023:12:16:03:03:11]: Info: System tray created
[2023:12:16:03:03:11]: Error: Failed to create session: Version mismatch between NvFBC and the X driver interface
[2023:12:16:03:03:11]: Error: Unable to initialize capture method
[2023:12:16:03:03:11]: Error: Platform failed to initialize
[2023:12:16:03:03:11]: Info: Trying encoder [nvenc]
[2023:12:16:03:03:12]: Info: Encoder [nvenc] failed
[2023:12:16:03:03:12]: Error: Couldn't find any working encoder matching [nvenc]
[2023:12:16:03:03:12]: Info: // Testing for available encoders, this may generate errors. You can safely ignore those errors. //
[2023:12:16:03:03:12]: Info: Trying encoder [vaapi]
[2023:12:16:03:03:12]: Info: Encoder [vaapi] failed
[2023:12:16:03:03:12]: Info: Trying encoder [software]
[2023:12:16:03:03:13]: Info: Encoder [software] failed
[2023:12:16:03:03:13]: Fatal: Couldn't find any working encoder
[2023:12:16:03:03:13]: Error: Video failed to find working encoder
```

**Comment by KelvinGuy1** [comment]
Update: Switching back to x11 does not resolve the issue.

**Comment by Nitrousoxide** [comment]
I have this issue on Fedora as well with the flatpak.  Encoders are installed as flatpak runtimes and accessible to other streaming/recording flatpaks like OBS.

**Comment by AAdmiral56** [comment]
For me on Pop!_OS it works on X11 but not wayland. OBS doesnt have any encoders available either except CPU, and sunshine doesnt even have that on wayland.

**Comment by Nitrousoxide** [comment]
hmm, yeah X11 finds vaapi with the Sunshine flatpak on Fedora 39.  Nothing under wayland though.

**Comment by AAdmiral56** [comment]
Oh a note: I am on a Vega 64, so it appears the issue is vendor agnostic.

**Comment by Nitrousoxide** [comment]
I have this issue on amd gpu and I've seen other people with nvidia gpus's complain about it on the discord, so I think it's also not dependent on hardware issue.  I'm almost certain it's a documentation issue or a bug that needs the devs to resolve.

**Comment by SimonSchwendele** [comment]
You'll see it in the CI
https://github.com/LizardByte/Sunshine/blob/nightly/docker/fedora-39.dockerfile#L70

nowadays most of us are already on gcc13 while cuda is only supported until gcc12.3.
Cuda is the piece of software that is required by sunshine to allow hardware acceleration on nvidia gpus.

For the time being we might've to wait until cuda gets updated.
https://docs.nvidia.com/cuda/cuda-installation-guide-linux/index.html#system-requirements

**Comment by ReenigneArcher** [comment]
@SimonSchwendele that's a Fedora issue. It doesn't affect Debian or Flatpak.

**Comment by SimonSchwendele** [comment]
> @SimonSchwendele that's a Fedora issue. It doesn't affect Debian or Flatpak.

aah damn, the reply didn't work.
That should've replied to @Nitrousoxide

**Comment by Nitrousoxide** [comment]
> @SimonSchwendele that's a Fedora issue. It doesn't affect Debian or Flatpak.

I have this issue in flatpak using Wayland. Though with an AMD GPU. But regardless it should still find a software encoder at the very least regardless of which packaging utility and compositor is used.  Doing literally nothing else other than logging out from a Wayland session into an X11 session will let the flatpak version properly screencast as it is able to find encoders.

There's something clearly wrong with either its settings for Wayland or the instructions that people are given in the docs.

**Comment by pearagone** [comment]
[My comment](https://github.com/LizardByte/Sunshine/issues/1631#issuecomment-1869209874) on #1631 may be relevant, I am also using Wayland on Fedora.

**Comment by ReenigneArcher** [comment]
Duplicate of #1631

**Comment by gittykrabs** [comment]
Exact same problem for me regardless of what Linux distro I choose. Switching to Xorg results in the Fatal: Couldn't find any working encoder message going away however, it still doesn't work. Only in software mode which is impossible to use. In my case, an Intel N100 based machine. Haven't been able to find a fix as yet.

**Comment by SimonSchwendele** [comment]
> Exact same problem for me regardless of what Linux distro I choose. Switching to Xorg results in the **Fatal: Couldn't find any working encoder** message going away however, it still doesn't work. Only in software mode which is impossible to use. In my case, an Intel N100 based machine. Haven't been able to find a fix as yet.

huh... this issue is just an nvidia problem.
What das `vainfo` tell you ?
Maybe you need to test some stuff.
`LIBVA_DRIVER_NAME=i965 vainfo` or `LIBVA_DRIVER_NAME=iHD vainfo`.
If both yield errors then maybe your chip can't do hardware acceleration.

Either way till nvidia eventually releases cuda13 I am too using the intel vaapi ( i9-13900H ), offload the applications and deal with the performance impact.

Anyways please check that your hardware accelerations works at all, this should not be a sunshine issue

**Comment by gittykrabs** [comment]
Thanks for the reply. Is it true thag the Quicksync Encoder option is only on Windows?

I'll have a look at your suggestions when I'm at my PC. Many thanks again!

**Comment by SimonSchwendele** [comment]
> I wonder if i even works on wayland (nvidia) at all? I installed the gigantic cuda package and it still falls back on CPU.

![Image](https://github.com/user-attachments/assets/04d3fd49-f929-44bc-8517-7c568e1e7e9e)

No Problem here.
I'm on Fedora using the [copr feed of matte-schwartz](https://copr.fedorainfracloud.org/coprs/matte-schwartz/sunshine/)

**Comment by shelterx** [comment]
Sure.. It finds nvenc for me too but is it really used when you start a stream? Because for me it says it's falling back to CPU.

**Comment by SimonSchwendele** [comment]
> Matte's copr is no longer updated/maintained.

Oh thanks didn't see this.
I started using it during the time cuda was unsupported due to a too recent gcc

**Comment by shelterx** [comment]
> cuda was unsupported due to a too recent gcc

I'm wondering if that's the case here, cuda seems to be using gcc 13 libs and the rest of the system uses gcc 14?

**Comment by shelterx** [comment]
I compiled it myself from the AUR using GCC13 and it worked (I think? See below).

The PKGBUILD file needed some adjustment tho'. There was a dependency for 'ffnvcodec-headers' but those were already on the system so I removed it, I also had to enable cuda in the PKGBUILD file.
```
[2025-01-23 08:59:38.228]: Info: Creating encoder [hevc_nvenc]
[2025-01-23 08:59:38.228]: Info: Color coding: SDR (Rec. 601)
[2025-01-23 08:59:38.228]: Info: Color depth: 8-bit
[2025-01-23 08:59:38.228]: Info: Color range: MPEG

```

**Comment by ReenigneArcher** [comment]
If you're on arch, we have precompiled versions in our pacman-repo

**Comment by shelterx** [comment]
@ReenigneArcher Sorry if I'm reading things wrong but I don't think the instructions here will work, the URLs are invalid?
https://github.com/LizardByte/pacman-repo/

**Comment by Ducduong1811** [comment]
The solution that worked for me was:

1. I was installing Nvidia Driver according to this [answer](https://www.reddit.com/r/Ubuntu/comments/1cy14b3/comment/lno1nek/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button). 
2. After installing, it tells me that Sunshine requires miniupnpc package to work. I then `sudo apt install miniupnpc`, run `sunshine` again and it works perfectly

TL;DR:  Seems like `sudo apt install miniupnpc` is the key here

**Comment by ReenigneArcher** [comment]
> The solution that worked for me was:
> 
>     1. I was installing Nvidia Driver according to this [answer](https://www.reddit.com/r/Ubuntu/comments/1cy14b3/comment/lno1nek/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button).
> 
>     2. After installing, it tells me that Sunshine requires miniupnpc package to work. I then `sudo apt install miniupnpc`, run `sunshine` again and it works perfectly
> 
> 
> TL;DR: Seems like `sudo apt install miniupnpc` is the key here

`miniupnpc` is already listed as a run time dependency for the deb packages. => https://github.com/LizardByte/Sunshine/blob/fe07501820b0bdd734321f9716b9dd79d3ca886f/cmake/packaging/linux.cmake#L86
